staging/src/k8s.io/apiserver/pkg/storage/cacher/ready.go

```diff
@@ -30,26 +31,43 @@ const (
 )
 
 // ready is a three state condition variable that blocks until is Ready if is not Stopped.
-// Its initial state is Pending.
+// Its initial state is Pending and its state machine diagram is as follow.
+//
+// Pending <------> Ready -----> Stopped
+//
+//	|                           ^
+//	└---------------------------┘
 type ready struct {
-	state status
-	c     *sync.Cond
+	state       status        // represent the state of the variable
+	lock        sync.RWMutex  // protect the state variable
+	restartLock sync.Mutex    // protect the transition from ready to pending where the channel is recreated
+	waitCh      chan struct{} // blocks until is ready or stopped
 }
 
 func newReady() *ready {
 	return &ready{
-		c:     sync.NewCond(&sync.RWMutex{}),
-		state: Pending,
+		waitCh: make(chan struct{}),
+		state:  Pending,
 	}
 }
 
+// done close the channel once the state is Ready or Stopped
+func (r *ready) done() chan struct{} {
+	r.restartLock.Lock()
+	defer r.restartLock.Unlock()
+	return r.waitCh
+}
+
 // wait blocks until it is Ready or Stopped, it returns an error if is Stopped.
-func (r *ready) wait() error {
-	r.c.L.Lock()
-	defer r.c.L.Unlock()
-	for r.state == Pending {
-		r.c.Wait()
+func (r *ready) wait(ctx context.Context) error {
+	select {
+	case <-ctx.Done():
+		return ctx.Err()
+	case <-r.done():
 	}
+
+	r.lock.RLock()
```

Comment:
I have one more concern here - given that those are now separate critical sections the following seems possible:

0) state is Pending
1) caller A is calling wait()
2) in the meantime caller B is calling set(true)
3) this unblocks caller A from waiting on r.done()
4) in the meantime caller B is calling set(false)
5) when caller A continues processing, the state is Pending again and we return an error("unexpected apiserver cache state)

In the context of how we're using `ready` it may not be the end of the world, but it's pretty misleading...

Reply:
you have Xrays for these things , you are right 😄 

Added a test and handled the PEnding state on the wait function

Reply:
> you have Xrays for these things , you are right

:D
I missed such things in the past and I finally learnt that :)

Reply:
@wojtek-t check new commit b7beea396dcc8291ac107b97035ad82db134102a , it moves back to pending to avoid having some process waiting and others keep going